distributed/scheduler.py

```diff
@@ -453,8 +453,13 @@ class WorkerState:
     #: Arbitrary additional metadata to be added to :meth:`~WorkerState.identity`
     extra: dict[str, Any]
 
+    # Used to distinguish otherwise equal WorkerState objects (e.g. if the worker was created with same addresses)
+    _instance_id: int
```

Comment:
I'd kinda lean towards comparing by ID. There _should_ always be exactly one WorkerState instance per Worker (process).

Using the ID as the hash I think would even be safe, because obviously you can't have two objects with the same ID as the same time (so you could never have a collision where a dict contains an old instance, and a new instance re-used that ID). If you're storing the actual `hash(ws)` values somewhere and using them for a manual equality check, then this could be incorrect, but that also seems like something you just should never do.

I'm fine with leaving the counter just as a guarantee that the hashes are unique as well, but it may be unnecessary.

I'd be inclined to remove the `__eq__` method, since I believe WorkerState instances should not ever be compared via `==`, since there should never be 2 different instances that refer to the same worker.

Therefore we should remove `__hash__` as well:

> User-defined classes have [`__eq__()`](https://docs.python.org/3/reference/datamodel.html#object.__eq__) and [`__hash__()`](https://docs.python.org/3/reference/datamodel.html#object.__hash__) methods by default; with them, all objects compare unequal (except with themselves) and `x.__hash__()` returns an appropriate value such that `x == y` implies both that x is y and `hash(x) == hash(y)`.
https://docs.python.org/3/reference/datamodel.html#object.__hash__

The objects will still be hashable in that case. They're only unhashable if they define `__eq__` but not `__hash__`.